hubspot/plugin.py

```diff
@@ -0,0 +1,264 @@
+
+    Args:
+        contacts: A list of HubSpotContactUpdate objects to be updated.
+
+    Returns:
+        A list of strings, where each string is the HubSpot ID of an updated contact.
+    """
+    url = "https://api.hubapi.com/crm/v3/objects/contacts/batch/update"
+
+    # Prepare the payload from the contacts list
+    contacts_payload = []
+    for contact in contacts:
+        contact_data = {
+            "id": contact.id,  # Contact ID is required for updating
+            "properties": contact.updated_properties,
+        }
+        contacts_payload.append(contact_data)
+
+    payload = {"inputs": contacts_payload}
+
+    with httpx.Client(
+        transport=AugmentedTransport(actions_v0.authenticated_request_hubspot),
+    ) as client:
+        response = client.post(url, json=payload)
+        response.raise_for_status()
+        data = response.json()
+
+    # Extract and return the IDs of the updated contacts
+    updated_contact_ids = [result["id"] for result in data["results"]]
+    return updated_contact_ids
+
+
+def search_contacts(
+    search_criteria: Dict[str, str], additional_properties: Sequence[str] | None = None
+) -> List[HubSpotContact]:
+    """
+    Searches for HubSpot contacts based on various criteria.
+
+    Parameters:
+        search_criteria: A dictionary where keys are the property names (e.g.,
+          "firstname", "email") and values are the search values for those properties.
+        additional_properties: A sequence of property names to fetch from found
+            contacts. If present, the corresponding values will be provided in the
+            HubSpotContactProperties additional_properties field. Standard hubspot
+            properties are available, but users must know the names of custom properties
+            if they are to be found.
+
+    Returns:
+        List[HubSpotContact]: A list of HubSpotContact objects matching the search
+            criteria.
+    """
+    url = "https://api.hubapi.com/crm/v3/objects/contacts/search"
+
+    # Construct the filters based on the search criteria
+    filters = []
+    for property_name, value in search_criteria.items():
+        filters.append(
+            {"propertyName": property_name, "operator": "EQ", "value": value}
+        )
+
+    properties = ["firstname", "lastname", "email", "lastmodifieddate"]
+    if additional_properties:
+        properties = list(set(properties + list(additional_properties)))
+
+    # Prepare the request body with the filters
+    payload = {"filterGroups": [{"filters": filters}], "properties": properties}
+
+    contacts = []
+    with httpx.Client(
+        transport=AugmentedTransport(actions_v0.authenticated_request_hubspot),
+    ) as client:
+        response = client.post(url, json=payload)
+        response.raise_for_status()
+        data = response.json()
+
+    for item in data.get("results", []):
+        properties = item.get("properties", {})
+        addl_props = {}
```

Comment:
Perhaps `property_values` and `additional_property_values` to avoid the "shadowing" and change in abbreviation?